### ISSUE TITLE
Make sure functions with two return statements return objects of the same type

### ISSUE THREAD

**Issue by omarandlorraine** [PyCQA/pylint#6580]
### Current problem

I'm trying to untangle a function that's several hundreds lines of unmanageable spaghetti code. The type of the returned object varies with global state and some of the function's many parameters.

I don't know if a function returning both integer and None is okay, but I think pylint should definitely b0rk at returning sometimes a 5-tuple and sometimes a 7-tuple.

### Desired solution

I want to have a report that warns me that a function does not always return things of the same type. 

Maybe if the function sometimes returns None it's okay, or if type A inherits from type B. More research/reasoning needed here maybe.

### Additional context

I don't see a corresponding PEP guideline, maybe I missed it.

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. On the title alone I would close as won't fix because while it's interesting, it's out of scope for pylint. It's what mypy or pyright do, and they do it well.

> Maybe if the function sometimes returns None it's okay,

This is ``Optional[x]`` or ``x | None``.it can be checked by mypy already.

> I think pylint should definitely b0rk at returning sometimes a 5-tuple and sometimes a 7-tuple.

This part though, is in scope for pylint, it could be a check for ``too-many-object-in-return`` or something close.

**Comment by cdce8p** [comment]
> Thank you for opening the issue. On the title alone I would close as won't fix because while it's interesting, it's out of scope for pylint. It's what mypy or pyright do, and they do it well.

I agree. That's something type checkers can already do quite well.

> > I think pylint should definitely b0rk at returning sometimes a 5-tuple and sometimes a 7-tuple.
> 
> This part though, is in scope for pylint, it could be a check for `too-many-object-in-return` or something close.

This is also something type checkers can do if the function is annotated correctly. E.g. both mypy and pyright would reject a tuple with 7 int items if the return type is `tuple[int, int, int, int, int]`.

Regarding a `too-many-object-in-return` error, not sure how useful that would be overall. Sure we can check all return statements, but not only is that expensive, we might not even catch all instances if for example a comprehension is used. In the end it could also be the desired output or kept for backwards compatibility reasons. The false-positive rate would be high for not much usefulness. The only thing we could do is to suggest the use of `NamedTuple` as return type but then again I'm not sure we should do that either.

**Comment by jacobtylerwalls** [comment]
Marc makes a good point that precise type annotation for tuples means this is better handled up and down with a type checker. +1 wontfix

**Comment by Pierre-Sassoulas** [comment]
I misread what @omarandlorraine was saying, I was thinking of a refactoring check for "big typing" like more than 7 elements for example if you have this function:
```python
def xombulate() -> Dict[
    str,
    Union[
        None,
        str,
        bool,
        int,
        Pattern[str],
        Iterable[Union[str, int, Pattern[str]]],
        Type["_CallbackAction"],
        Callable[[Any], Any],
        Callable[[Any, Any, Any, Any], Any],
    ],
]:
    ...
```
Then maybe it's time to create a class or a typing alias. But it's another proposition as this one is about typing mismatch, I'm going to close as won't fix.
